js-pkg/client/src/provider.ts

```diff
@@ -143,21 +154,83 @@ export class YSweetProvider {
     this.awareness.on('update', this.handleAwarenessUpdate.bind(this))
     this.WebSocketPolyfill = extraOptions.WebSocketPolyfill || WebSocket
 
+    this.online = this.online.bind(this)
+    this.offline = this.offline.bind(this)
+    if (typeof window !== 'undefined') {
+      window.addEventListener('offline', this.offline)
+      window.addEventListener('online', this.online)
+    }
+
     doc.on('update', this.update.bind(this))
 
     if (extraOptions.connect !== false) {
       this.connect()
     }
   }
 
+  private offline() {
+    // When the browser indicates that we are offline, we immediately
+    // probe the connection status.
+    // This accelerates the process of discovering we are offline, but
+    // doesn't mean we entirely true the browser, since it can be wrong
+    // (e.g. in the case that the connection is over localhost).
+    this.checkSync()
+    this.setConnectionTimeout()
+  }
+
+  private online() {
+    if (this.reconnectSleeper) {
+      this.reconnectSleeper.wake()
+    }
+  }
+
+  private clearHeartbeat() {
+    if (this.heartbeatHandle) {
+      clearTimeout(this.heartbeatHandle)
+      this.heartbeatHandle = null
+    }
+  }
+
+  private setHeartbeat() {
+    if (this.heartbeatHandle) {
+      return
+    }
+    this.heartbeatHandle = setTimeout(() => {
+      this.checkSync()
+      this.setConnectionTimeout()
+      this.heartbeatHandle = null
+    }, MAX_TIMEOUT_BETWEEN_HEARTBEATS)
+  }
+
+  private clearConnectionTimeout() {
+    if (this.connectionTimeoutHandle) {
+      clearTimeout(this.connectionTimeoutHandle)
+      this.connectionTimeoutHandle = null
+    }
+  }
+
+  private setConnectionTimeout() {
+    if (this.connectionTimeoutHandle) {
+      return
+    }
+    this.connectionTimeoutHandle = setTimeout(() => {
+      if (this.websocket) {
+        this.websocket.close()
+        this.setStatus(STATUS_ERROR)
+        this.connect()
+      }
+      this.connectionTimeoutHandle = null
+    }, MAX_TIMEOUT_WITHOUT_RECEIVING_HEARTBEAT)
+  }
+
   private send(message: Uint8Array) {
     if (this.websocket?.readyState === this.WebSocketPolyfill.OPEN) {
       this.websocket.send(message)
     }
   }
 
   private updateSyncedState() {
-    let hasLocalChanges = this.lastSyncAcked !== this.lastSyncSent
+    let hasLocalChanges = this.ackedVersion !== this.localVersion
     if (hasLocalChanges === this.hasLocalChanges) {
       return
     }
```

Comment:
I like that, done. Unfortunately it adds a bit of complexity at the former call sites because we want to emit if and only if the `hasLocalChanges` value would change, but I do think it's an improvement overall.